system/jlib/jfile.cpp

```diff
@@ -7106,3 +7109,230 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
+                    filename = event->name;
+                else
+                {
+                    /* if watching a file, lookup filename from event handle
+                     * linear search, not worth hash table, not expected to be that many.
+                     */
+
+                    CriticalBlock b(crit);
+                    for (const auto &item: monitoredFiles)
+                    {
+                        if (event->wd == item->watchFd)
+                            filename = item->filename;
+                    }
+                }
+                /* NB: in this impl. inotify_event flags are same as FileWatchEvents bit masks
+                 * If other implementations differ, they will need mapping to FileWatchEvents
+                 */
+                FileWatchEvents events = static_cast<FileWatchEvents>(event->mask);
+
+                callback(filename.c_str(), events);
+
+                ptr += sizeof(struct inotify_event) + event->len;
+            }
+            while (ptr < (buf + len));
+        }
+    }
+    void checkClose(int h)
+    {
+        if (-1 != h)
+            close(h);
+    }
+public:
+    CFileEventWatcher(FileWatchFunc _callback) : callback(_callback), threaded("CFileEventWatcher", this)
+    {
+        inotifyFd = inotify_init1(IN_NONBLOCK);
+        if (-1 == inotifyFd)
+            throw makeStringException(-1, "CFileEventWatcher - inotify_init1");
+        if (pipe(pipefd))
+            throw makeErrnoException(-1, "CFileEventWatcher - create pipe failed");
+        for (unsigned i=0; i<2; i++)
+        {
+            int flags = fcntl(pipefd[i], F_GETFL, 0);
+            if (-1 != flags)
+            {
+                flags |= O_NONBLOCK;
+                fcntl(pipefd[i], F_SETFL, flags);
+            }
+            flags = fcntl(pipefd[i], F_GETFD, 0);
+            if (-1 != flags)
+            {
+                flags |= FD_CLOEXEC;
+                fcntl(pipefd[i], F_SETFD, flags);
+            }
+        }
+
+        toPoll[0].fd = pipefd[0];
+        toPoll[0].events = POLLIN;
+        toPoll[1].fd = inotifyFd;
+        toPoll[1].events = POLLIN;
+    }
+    ~CFileEventWatcher()
+    {
+        stop();
+        checkClose(inotifyFd);
+        checkClose(pipefd[0]);
+        checkClose(pipefd[1]);
+    }
+    virtual void start() override
+    {
+        if (stopped)
+        {
+            stopped = false;
+            threaded.start();
+        }
+    }
+    virtual void stop() override
+    {
+        if (stopped)
+            return;
+        stopped = true;
+
+        const byte c = 0;
+        if (write(pipefd[1], &c, 1) != 1)
+        {
+            Owned<IException> e = makeErrnoException("CFileEventWatcher::stop() - write -> pipefd[1]");
+            EXCLOG(e);
+        }
+        threaded.join();
+    }
+    virtual bool add(const char *filename, FileWatchEvents events) override
+    {
+        CriticalBlock b(crit);
+        for (const auto &item: monitoredFiles)
```

Comment:
Hash table?

Reply:
need 2 (lookup by fd and filename), have changed to use ht's.